src/sdam/monitor.rs

```diff
@@ -114,10 +130,28 @@ fn check_server(
         }
     }
 
-    ServerDescription::new(address, Some(is_master(conn)))
+    ServerDescription::new(address, Some(is_master(conn, make_connection)))
+}
+
+fn is_master(
+    conn: &mut Connection,
+    make_connection: impl Fn() -> Connection,
+) -> Result<IsMasterReply> {
+    let result = is_master_inner(conn);
+
+    if result
+        .as_ref()
+        .err()
+        .map(|e| e.kind.is_network_error())
+        .unwrap_or(false)
+    {
+        std::mem::replace(conn, make_connection());
+    }
+
+    result
 }
 
-fn is_master(conn: &mut Connection) -> Result<IsMasterReply> {
+fn is_master_inner(conn: &mut Connection) -> Result<IsMasterReply> {
```

Comment:
this function is only called above and it's short enough that I think it makes sense to just include it inline in `is_master`.

Reply:
I separated it so that we could use the `?` operator in the body. The idea is that we want to short-circuit doing the rest of this method in any of the numerous places errors could occur, but we don't want to exit from `is_master` itself because we want to replace the connection if an error occurred.

Reply:
Makes sense. Now that I think about it though, we should just call `make_connection` in the error handling portion of `check_server`, rather than inside of `is_master`. Connection reestablishment isn't really a concern for an isMaster helper anyways. This would remove the need for an `is_master_inner`.

Also, the spec says that we should only be doing some of this logic in response to connection errors and timeout errors, but we seem to just be doing it for all errors.

Reply:
We call `is_master` twice from `check_server`, which is why it's abstracted into a function. I think it's cleaner not to duplicate the code.

Reply:
> Also, the spec says that we should only be doing some of this logic in response to connection errors and timeout errors, but we seem to just be doing it for all errors.

The code checks to see if the error is a network error (i.e. an I/O error). If it's not, the connection isn't replaced.

Reply:
I was referring to the error checking in `check_server`. Another benefit of moving the connection re-establishment there is that the error handling of network errors is consolidated, as it is in the SDAM spec psuedocode.